conquest.man
.\" $Id$
.TH "conquest" 6 "" ""
.SH NAME
conquest \- Conquest game clients
.SH SYNOPSIS
.PP
\fBconquest\fP is a Freeglut OpenGL based client that is used to connect to a conquest serverto play.
.PP
\fBconquest\fP [OPTIONS]...
.SH DESCRIPTION
.PP
Conquest is a multi-player, real-time space combat game.  The object
of the game is twofold. The short-range goal is to accumulate "kills"
by shooting down enemy players.  You get one kill point for each enemy
ship shot down, plus some extra if the enemy had kills too. The major
weapon used to shoot down ships is the photon torpedo.
.PP
The long-range goal is to conquer the universe for your team by taking
every \fIcore\fP planet. You take planets by killing off the enemy's armies via
bombardment, and then beaming your team's armies down. When all the
planets have been taken, the game ends, a new game begins, and the
player who actually took the last planet gets his/her name up in
lights.
.PP
The first time \fBconquest\fP is run, a directory
called ~/.conquest/ will be created in your home directory.  This
directory cantains your 'user options' data as well as the logfile.
Additionally, texturesrc (.trc) files and
texture image data (in ~/.conquest/img/) can be stored there. See
CONQINIT.TXT for more information on what \fItexturesrc\fP and
\fItexture image files\fP are.

.SH "OPTIONS"
.PP
If the \fI\-s\fP or \fI\-m\fP options are
not specified, the client will try to connect to a server running on
your local host at port 1701.
.TP 8
.B \-B
Benchmark mode.  When this option is specified, and playback of a
previously recorded game is specified with \fI\-P\fP, then the default
playback speed will be set to the highest speed possible (no frame
delay).

It's not really much of a 'benckmark' yet as it does not currently
emit any rendering statistics on exit.
.TP 8
.B \-f
Run in fullscreen mode.
.TP 8
.B \-g <geometry> 
Specify initial window width/height.  OpenGL client only.  Format is
WxH (ex: 1024x768). The default is currently 1280x720.
.TP 8
.B \-m  
Query the metaserver and provide a list of servers to choose from.
.TP 8
.B \-M <metaserver> 
Specify an alternate metaserver.  The default is
\fIconquest.radscan.com\fP. 
.TP 8
.B \-r <record file> 
Record the game to <record file>.
.TP 8
.B \-s <server[:port]> 
Connect to server running on \fIserver\fP port \fIport\fP.  The default
port is 1701.
.TP 8
.B \-P <record file> 
Play back a previously recorded game.
.TP 8
.B \-u
Disable UDP to/from the server and use TCP only.  The default is to
try to negotiate UDP and use it for certain conquest protocol
packets.  If you have problems with a server \- for example, you
always wind up stuck in the middle of Murisak when you enter the game,
try disabling UDP with this option.

.TP 8
.B \-S
Disable sound support.  If \fBconquest\fP was compiled with sound
support, then this option will disable the playing of all music and
effects.

.SH "IMPORTANT COMMANDS"
.PP
This section lists the basic commands of Conquest.  The next section
lists the more advanced / obscure commands, and at the end of this
document there is a summary list.  The list is also available in the game
via the \fBh\fP command.
.TP 8
.B 0\-9,= 
To set your warp factor, just type a digit, or \fB=\fP for your ship's
maximum warp. The ships teams have different top speeds: 8 for Cruisers, 9
for Destroyers, and 10 for Scouts.
.TP 8
.B -,+
Use \fB+\fP to raise your energy shields, \fB-\fP to lower them.  Shields
protect you from enemy weapons; however, it takes more fuel and heat
to fly with them up.
.TP 8
.B k 
The \fBk\fP command is used to set your course.  You can specify your
course a number of different ways: in degrees (0=east,90=north), via
the arrow keys (see below), a ship number (precede it with an 's'), or
a planet name.  If you set course for a planet, you have the option of
locking on to it by terminating the line with a TAB instead of a
carriage return - then you will orbit automatically when you get
there.
.TP 8
.B p,P 
These two commands fire photon torpedoes.  Lower case \fBp\fP fires one at
a time, and is less useful. Upper case \fBP\fP fires a burst of three.
When you are in the heat of battle, you will want to fire every torp
you have got (nine), so you will almost always use \fBP\fP. You can
specify the direction to fire using either degrees or the arrow keys.
.TP 8
.B *,d,D 
Lower case \fBd\fP or \fB*\fP detonates near-by enemy torps before they get
close enough to damage you.  Use this command *alot*.  Upper case \fBD\fP
detonates your own torps, for instance in case you missed.  Since you
can only have nine torps out at a time, this can be useful.
.TP 8
.B f,F
These two commands let you fire phasers.  Lower case \fBf\fP is used to
fire one phaser blast. It takes about one second to fire
phasers. Upper case \fBF\fP fires phasers again in the same direction.
.TP 8
.B B
To capture enemy planets, you first have to get rid of the enemy
armies by bombarding them.  Orbit the planet, type \fBB\fP, confirm with a
TAB, and then watch your damage closely. You will continue to bombard
until you type another command.  When your damage gets up above 60 or
70 percent, set a warp factor and get out fast.

You cannot bomb a planet down to less than 3 armies.  Once you have
reduced a planet's army count to 3, you must beam down your own armies
in order to kill the remaining armies and conquer the planet.
.TP 8
.B R
After a bombing run, you can repair your damage three times faster by
going into repair mode.  While repairing, your shields are down and
you can't move or fire.  Repairing is useful any time you are damaged.
If you are in orbit about a friendly planet when you repair, you will
repair even faster.
.TP 8
.B b
Once you have blasted a planet down to three armies, you can beam your
own armies down to do hand-to-hand combat and replace them.
.TP 8
.B m
Use the \fBm\fP command to send messages. You can send to single ships
(using the ship number), entire teams, (first letter of team name) or
everyone ("all"). You can even send messages to GOD, but don't expect
an answer.  Messages to GOD will be seen be any conquest operators
(CO's) who happen to be in the game.  

This command is useful for bothering the experts and asking them
annoying questions.  It is also useful for trading insults.  If you
type RETURN or TAB as the target of a message, the message will be
sent to the last person(s) you sent a message to.
.TP 8
.B i,TAB,RETURN
You can get information on planets and ships with the \fBi\fP command.
Pressing RETURN will show you the last information you asked for.
Pressing TAB will show you the 'next' last information you asked for.
.TP 8
.B Q 
The only way out besides getting killed is to self-destruct. There is
a 15-second time delay because it is not honorable to quit in the heat
of battle.  If the server has the \fIVacant\fP flag enabled, then you
can use the Control-Backslash key to quit immediately, leaving your
ship intact.  If you decide to go \fIVacant\fP, be careful where you
leave your ship. :)
.TP 8
.B /
To see which other people are currently playing Conquest, use the \fB/\fP
command.
.TP 8
.B h
This command gives you a short command summary (such as the one that
is found at the end of this document).
.SH "LESS IMPORTANT COMMANDS"
.PP
.TP 8
.B A
Type \fBA\fP to change your weapons/engines power allocations.  You can
either enter a new weapons allocation percentage or else type TAB to
toggle between the two extremes of 30/70 and 70/30. The power
allocation controls things like how powerful your weapons are, how
efficient your engines are, etc.
.TP 8
.B C
Cloaking device. You can activate the cloaking device by typing \fBC\fP
followed by a TAB.  While the cloaking device is in operation, you are
invisible, even to your teammates. However, the cloaker uses a LOT of
energy - so much that you can't use your weapons while it's on. A
second \fBC\fP will turn it off again.
.TP 8
.B E
Emergency distress signal. A quick way to let your team mates know
that you are in trouble is by typing \fBE\fP followed by a TAB.  Every
ship on your team (except those who are at war with your team) will
get a short message from you that includes information about your
fuel, shields, and ship damage.
.TP 8
.B H
History. This command displays the list of recent Conquest players.
.TP 8
.B O
Use this comment to enter the options menu.  From there you can
configure various things like your settings and your macros.
.TP 8
.B K
Attempt a coup. If all of your team's armies get killed, you may think
you are in deep fecal matter.  Not so - you can still get some back by
pulling a coup.  Try it.
.TP 8
.B L
List Messages.  Type \fBL\fP to review old messages that have been
sent to you.
.TP 8
.B M
Type \fBM\fP to toggle your display between the short range and long range
tactical displays.

When in the long range tactical display, you can use SPACE to toggle
the center of the map beteen your ship and the Murisak (the sun at the
center of the universe).
.TP 8
.B N
Pseudonym.  Use this command to change your battle name.
.TP 8
.B o
Orbit. Use \fBo\fP to come into orbit around a planet (or a sun, if you
want a tan).  You cannot enter orbit if your speed exceeds warp 2.
.TP 8
.B r
Refit.  If the server has enabled the \fIRefit\fP flag, and you have
at least one kill, and you are orbiting a team owned planet, you can
use this option to switch to a new ship type.  Valid types are Scout
(great engines, weak weapons), Destroyer (good engines, good weapons),
and Cruiser (weak engines, great weapons).
.TP 8
.B S
Displays a list of secondary user statistics.
.TP 8
.B t
Tractor beams.  You can use these to tow another ship, even if it is
hostile.  This can be fun when you are cloaked :)
.TP 8
.B T
Team list. Displays a list of team statistics, and information on when
and who last conqered the universe.
.TP 8
.B u
Untractor. Disengage your tractor beams if you are towing another ship
or attempt to break free of another ships tractor beam if you are
being towed.
.TP 8
.B U
Displays a list of users and some primary statistics.
.TP 8
.B W
Type \fBW\fP to change your war and peace settings.  Note that there is a
10 second delay for declaring war on another team.  Also, if you do
something nasty to another team such as bomb one of their planets,
your war with that team becomes sticky and you may not declare peace
with them until you get a new ship (by dying).
.TP 8
.B ?
To get a list of planets, use the \fB?\fP command.  Planets that must
be taken in order to conquer the universe are marked with a \fI+\fP sign.
.TP 8
.B ^L
When you type \fB^L\fP, the screen is cleared and redrawn.
.SH "DIRECTION KEYS"
.PP
The easiest and quickest way to change direction is with the direction
keys or the keypad if your terminal definition supports it.  However,
if you are prompted to enter a direction, for example to fire phasers,
or set course, you can use the letters "qweadzxc" on the left-hand
side of your keyboard, called "direction keys" for historical
reasons. The keys are assumed to be arranged as follows (if you have a
Dvorak keyboard, you are in trouble):

                    Q W E
                     \\|/
                   A--+--D
                     /|\\
                    Z X C

You can use them singly, e.g. 'd'  would  be  0  degrees, 'q'
would  be  135,  etc.  You  can  also  use  them  in
combination: "ed" would be halfway between 'e'  and  'd',
which  is  22.5  degrees; "edd" is like 'e' + 'd' + 'd' /
3, or 15 degrees.
.SH "DESTINATION SEARCHING"
.PP
There are several special strings you can use when setting course or
getting information:
.TP 8
.B ne
nearest enemy 
.TP 8
.B ns
nearest ship
.TP 8
.B nts
nearest (friendly) team ship
.TP 8
.B np
nearest planet
.TP 8
.B nep
nearest enemy planet
.TP 8
.B nrp
nearest friendly (repair) planet
.TP 8
.B nfp
nearest friendly class-M (fuel) planet
.TP 8
.B nap
nearest planet with available armies (greater than 3 total armies)
.TP 8
.B ntp
nearest planet owned by your team
.TP 8
.B wp
weakest planet not owned by your team
.TP 8
.B hp
home planet for your team
.TP 8
.B sN
ship N (where N is a integer number)
.TP 8
.B time
time of day (info only)
.PP
All of these can be abbreviated to their shortest unique string. Also,
for the planets, you can type a number after the special string to
specify an army threshold; that is, planets with less than that number
of armies won't be considered.  For example, \fBna8\fP specifies the
nearest planet with 8 (eight) or more armies, \fBnf14\fP is the nearest
fuel planet with 14 or more armies, \fBnep1000\fP is the nearest
non-scanned planet.

You can find info out about the second nearest object by terminating
your information request with a TAB instead of a carriage return. For
example, to find out about the second nearest planet type
\fBinp<TAB>\fP. 

You can't use these when firing weapons - that would be too easy.

.SH "MACROS"
.PP
Macros are sequences of Conquest commands that are issued when a
Function Key (Fkey) or mouse button (OpenGL client only) is hit.  On
PC hardware, these are the F1-F12 keys.  With certain emulators the
Shifted F1-12 keys (F13-F24) are available too.

With the GL client, version 8.1.2f or better, support for assigning
macros to mouse buttons is also provided.  Modifiers like Alt,
Control, and Shift can also be used with the mouse buttons.  Up to 32
buttons are supported with any combination of the 3 modifiers (or no
modifiers) giving you a maximum of 256 assignable mouse macros.  If
you have a 32 button mouse that is :)

Mouse and Key macros are defined in your ~/conquest/conquestrc file.
Users can edit their macro keys from within Conquest using the
(O)ptions Menu.

Here is an example for the F1 function key

       macro_f1=dP\\r

Which makes my F1 key detonate enemy torps \fBd\fP, and fire a
spread of 3 torps in the last direction I fired \fBP\\r\fP.  

The mouse only works when playing the game (in the Cockpit) and the
cursor is within the viewer window in the GL client.

Three default mouse macros are provided to give you a taste, and will
be saved in your conquestrc file the first time you run an 8.1.2f or
better version of the client.  They are:

      mouse button 0 (left): Fire phaser at <angle>
      mouse button 1 (middle): Set course to <angle>
      mouse button 2 (right): Fire Torp at <angle>

Of course you can redefine these, as well as add others.  With mouse

HISTORY

         +--------------------------------------------------------+
         |  CCC    OOO   N   N   QQQ   U   U  EEEEE   SSSS  TTTTT |
         | C   C  O   O  NN  N  Q   Q  U   U  E      S        T   |
         | C      O   O  N N N  Q   Q  U   U  EEE     SSS     T   |
         | C   C  O   O  N  NN  Q  Q   U   U  E          S    T   |
         |  CCC    OOO   N   N   QQ Q   UUU   EEEEE  SSSS     T   |
         +--------------------------------------------------------+

                              Conquest HISTORY


8.5.1 (stable, windows only) 4/22/2010

  - add a call the WSAStartup() to conquestgl so socket i/o can
    *consistently* work :)

8.5 (stable) 04/20/2010

  - A new stable release, at long last.

  - Major highlights of this release, compared to the last stable
    release (8.4.1 09/13/2008):

    - General:

      - MAC OSX (Darwin) support.  Full sound and hardware accelerated
        OpenGL is supported.  Note, you will probably need to get SDL
        and SDL_mixer, as well as libogg and lib vorbis in order to
        build.  

      - Native Windows XP/Windows 7 support for the OpenGL client
        (conquestgl) using the MinGW cross-compiler enviroment.  Full
        sound and hardware accelerated OpenGL is supported.  A zip
        file will be available on sourceforge containing a
        pre-compiled conquestgl binary, and support files.  See
        README.Windows (or Conquest\INSTALL.TXT in the ZIP file) for
        more information on installtion and use.

      - a re-work of packet handling.  The original goal was to
        eventually support non-blocking socket IO, however some
        significant (and incompatible) protocol changes will be
        required to do this properly.  Next major release will tackle
        this. 

      - rework of texture and color representation internally.

      - remove telnet (-t) support from the clients.  Nobody runs
        telnet conquest games anymore.

      - switched to Mercurial (hg) for source code control.

        I will 'hg tag' each devel release.  For stable releases (like
        this one), a branch name will be created instead of a tag.

        Current development will always happen at the tip, in the
        'default' branch, and may therefore be unstable.

        Bug fixes and enhancements made to 8.5 will occur in the
        conquest-8.5 branch.

        A command like (all on one line):

          hg clone -u conquest-8.5 \
             http://conq.hg.sourceforge.net/hgroot/conq/conq conquest

          can be used to check out the stable branch for version 8.5.
          Omit the '-u conquest-8.5' if you want the current state of
          development.

    - Server fixes:

      - fix server bug that would cause vacant ships, when destroyed,
        to fail to explode properly, leaving some stationary torps
        behind.
      
      - correct a server problem with UDP keepalive packets.  During
        certain operations (like beaming, bombing, self destruct, etc)
        these packets could improperly terminate these activities on
        the server, leaving the client to believe these activities
        were still in progress, causing great angst in the land.

    - Client fixes (primarily the OpenGL client, conquestgl):

      - New Orion ship textures (thanks to blender and the fine
        contributors at ShareGC.com).  See the CREDITS file.

      - some changes to work better with wide-screen display formats

      - new ship shield textures, both for the icon-ship display and
        the main viewer.

      - in conquestgl, torps are no longer colored according to their
        'danger' level.  It's up to you, as the ship's Captain, to
        know when a torp may kill you.  When in doubt, raise
        shields and run like the coward you are. :)

      - rework of the icon-ship display:

        - now use a luminance variant of the main ship texture as the
          ico-ship rather than a seperate texture.

        - implement 'lamps' rather than the text (cloaking, repair,
          shields, etc)

        - make the ico-ship drawing area square to avoid stretching
          issues

        - use ship's torp texture for the torp-pips

        - use ship's phaser texture for the phaser recharge indicator

      - fix bug that could cause the use of undefined mouse macros to
        improperly terminate certain operations (beaming, bombing,
        etc). 

  - changes from the last devel (8.4.1i) release:

    - Fix up some documentation for Windows systems.  Conquestgl has
      been tested on Windows 7, seems to work great.

    - fix up stupidness in handling cInfo.remotehost data (caused
      invalid free()'s, depending on options specified).

    - Do not declare a crypt prototype on DARWIN systems

    - Move 'q' from beginning of line in conquest.man, lintian error
      reported by Jonathon (new debian maintainer)

    - Remove global.c, instantiate global.h and conf.h within all of
      the <main>.c files.  Needed by Darwin systems to avoid linker
      errors when building conqstrat and conqinit executables.

    - Add some icon pngs and a .ico for windows


  ## The following are the devel release notes for more detail.  Use
     'hg log -v' on the repository for even more detail if you care.

  8.4.1i (devel) 03/27/2010

  - unless there is a big problem, a 'stable' official release will be
    up next.  

  - Switch to using Mercurial (hg) for source code control.  I totally
    love it compared to svn.

  - MAC OS X (Darwin) support.  Note this support is still a little
    fuzzy, in regards to counquest group creation, etc.  I do not have
    direct access to a Darwin system, so this port was accomplished by
    back-and-forth with a friend who does.  Once you can get it built,
    OpenGL and sound is fully functional.

  - Windows (XP) native support via MinGW cross-compiler toolchain :)
    See README.Windows for installation instructions (or INSTALL.TXT

conquest.cxx
#include "c_defs.h"

/************************************************************************
 * The OpenGL client
 *
 * Copyright Jon Trulson under the MIT License. (See LICENSE).
 ***********************************************************************/


#include "conqdef.h"
#include "cb.h"

#include "conqutil.h"
#include "rndlb.h"
#include "conqunix.h"

#define NOEXTERN_CONTEXT
#include "context.h"

#define NOEXTERN_CONF
#include "conf.h"

#define NOEXTERN_GLOBALS
#include "global.h"

#include "color.h"
#include "record.h"
#include "ibuf.h"
#include "gldisplay.h"
#include "GL.h"
#include "conqnet.h"
#include "packet.h"
#include "client.h"
#include "clientlb.h"

#include "meta.h"

#include "nPlayBMenu.h"
#include "nConsvr.h"
#include "nMeta.h"

#include "conqinit.h"
#include "cqsound.h"
#include "hud.h"

#include "playback.h"

void catchSignals(void);
void handleSignal(int sig);
void conqend(void);

void printUsage()
{
    printf("Usage: conquest   [-m ][-s server[:port]] [-r recfile] [ -t ]\n");
    printf("                  [-f ][-g <geometry>] [ -u ]\n");
    printf("                  [ -M <metaserver> ] [ -B ]\n\n");
    printf("    -f               run in fullscreen mode\n");
    printf("    -g <geometry>    specify intial window width/height.\n");
    printf("                      Format is WxH (ex: 1024x768).\n");
    printf("                      Default is 1280x720.\n");
    printf("    -m               query the metaserver\n");
    printf("    -s server[:port] connect to <server> at <port>\n");
    printf("                      default: localhost:1701\n");
    printf("    -r recfile       Record game to <recfile>\n");
    printf("                      recfile will be in compressed format\n");
    printf("                      if conquest was compiled with libz\n");
    printf("                      support\n");

    printf("    -M metaserver    specify alternate <metaserver> to contact.\n");
    printf("                      default: %s\n", META_DFLT_SERVER);


    printf("    -P <cqr file>    Play back a Conquest recording (.cqr)\n");
    printf("    -B               Benchmark mode.  When playing back a recording,\n");
    printf("                      the default playback speed will be as fast as possible.\n");
    printf("    -u               do not attempt to use UDP from server.\n");
    printf("    -S               disable sound support.\n");
    printf("    -v               be more verbose.\n");


    return;
}

int getLocalhost(char *buf, int len)
{
    struct hostent *hp;

    gethostname ( buf, len );
    if ((hp = gethostbyname(buf)) == NULL)
    {
        if (buf)
            fprintf(stderr, "conquest: gethostbyname('%s'): cannot get localhost info.\n", buf);
        return FALSE;
    }

    return TRUE;
}

/* parse the geometry arg.  Ensure that dConf init w/h are only
 *  set if the geom arg was reasonably valid. Format WxH (ex: 1024x768)
 */
static void parseGeometry(char *geom)
{
    const int geoSize = 32;
    char geomcpy[geoSize];
    char *ch;
    int w, h;

    if (!geom || !*geom)
        return;

    memset((void *)geomcpy, 0, geoSize);
    utStrncpy(geomcpy, geom, geoSize);

    if ((ch = strchr(geomcpy, 'x')) == NULL)
        return;                     /* invalid */

    *ch = 0;
    ch++;

    if (!*ch)
        return;

    w = abs(atoi(geomcpy));
    h = abs(atoi(ch));

    if (!w || !h)
        return;

    /* set it up */
    dConf.initWidth = w;
    dConf.initHeight = h;

    return;
}

static int _cmpfile(void *cmp1, void *cmp2)
{
    char *str1 = *(char **)cmp1;
    char *str2 = *(char **)cmp2;
    return(strcmp(str1, str2));
}

static void _loadRCFiles(int type, const char *cqdir, const char *suffix)
{
    char filenm[PATH_MAX];
    DIR *dirp;
    struct dirent *direntp;
    char **filelist = NULL;
    char **tmplist = NULL;
    int numfiles = 0;
    int i;

    if (!cqdir || !suffix)
        return;


    if ((dirp = opendir(cqdir)))
    {
        while ((direntp = readdir(dirp)) != NULL)
        {                       /* here, we will just check out the .trc
                                   files */
            int len;
            int suff_len = strlen(suffix);

            if ((!strcmp(direntp->d_name, "..") ||
                 !strcmp(direntp->d_name, ".")))
                continue;

            len = strlen(direntp->d_name);
            if (len < suff_len)
                continue;

            if (!strncmp(&(direntp->d_name[len - suff_len]), suffix, suff_len))
            {                   /* found one */
                char *ptr;

                snprintf(filenm, PATH_MAX, "%s/%s", cqdir, direntp->d_name);

                if (!(ptr = strdup(filenm)))
                {
                    utLog("%s: Could not strdup %s, ignoring.",
                          __FUNCTION__,
                          filenm);

                }

                /* add it to the list */

                tmplist = (char **)realloc((void *)filelist,
                                           sizeof(char *) *
                                           (numfiles + 1));

                if (!tmplist)
                {
                    utLog("%s: Could not realloc %d files, ignoring '%s'",
                          __FUNCTION__,
                          numfiles + 1,
                          filenm);
                    if (ptr)
                        free(ptr);

                    continue;
                }

                filelist = tmplist;
                filelist[numfiles] = ptr;
                numfiles++;
            }
        }

        /* now sort it */
        if (!numfiles)
            return;

        qsort(filelist, numfiles, sizeof(char *),
              (int (*)(const void *, const void *))_cmpfile);

        /* print them out for testing */
        for(i=0; i <numfiles; i++)
        {
            cqiLoadRC(type, filelist[i], 1, 0);
            free(filelist[i]);
        }
        free(filelist);
        filelist = NULL;
        numfiles = 0;
    }

    return;
}

/* first load the main texturesrc and *.trc files.  Then look for and
   load any ~/.conquest/ *.trc files */
void loadTextureRCFiles()
{
    char cqdir[PATH_MAX];
    char *homevar;

    /* load the main texturesrc file first.  It's fatal if this fails. */
    if (!cqiLoadRC(CQI_FILE_TEXTURESRC, NULL, 1, 0))
        exit(1);

    /* now load any .trc files in there (CONQETC) */
    _loadRCFiles(CQI_FILE_TEXTURESRC_ADD, utGetPath(CONQETC), ".trc");

    /* now load any in the users own ~/.conquest/ dir */
    if ((homevar = getenv(CQ_USERHOMEDIR)) == NULL)
        return;

    snprintf(cqdir, PATH_MAX, "%s/%s",
             homevar, CQ_USERCONFDIR);

    _loadRCFiles(CQI_FILE_TEXTURESRC_ADD, cqdir, ".trc");

    return;
}

void loadSoundRCFiles()
{
    char cqdir[PATH_MAX];
    char *homevar;

    /* load the main sound file first */
    cqiLoadRC(CQI_FILE_SOUNDRC, NULL, 1, 0);

    /* now load any .src files in there (CONQETC) */
    _loadRCFiles(CQI_FILE_SOUNDRC_ADD, utGetPath(CONQETC), ".src");

    /* now load any in the users own ~/.conquest/ dir */
    if ((homevar = getenv(CQ_USERHOMEDIR)) == NULL)
        return;

    snprintf(cqdir, PATH_MAX, "%s/%s",
             homevar, CQ_USERCONFDIR);

    _loadRCFiles(CQI_FILE_SOUNDRC_ADD, cqdir, ".src");

    return;
}


/*  conquest - main program */
int main(int argc, char *argv[])
{
    int i;
    char *ch;
    int wantMetaList = FALSE;     /* wants to see a list from metaserver */
    int serveropt = FALSE;        /* specified a server with '-s' */
    int dosound = TRUE;
#if defined(MINGW)
    WSADATA wsaData;
    int rv;
#endif


    /* tell the packet routines that we are a client */
    pktSetClientMode(TRUE);
    pktSetClientProtocolVersion(PROTOCOL_VERSION);

    Context.entship = FALSE;
    Context.recmode = RECMODE_OFF;
    Context.updsec = 10;		/* dflt - 10/sec */
    Context.msgrand = time(0);

    utStrncpy(cInfo.metaServer, META_DFLT_SERVER, MAXHOSTNAME);
    cInfo.sock = -1;
    cInfo.usock = -1;
    cInfo.doUDP = FALSE;
    cInfo.tryUDP = TRUE;
    cInfo.state = CLT_STATE_PREINIT;
    cInfo.serverDead = TRUE;
    cInfo.isLoggedIn = FALSE;
    cInfo.remoteport = CN_DFLT_PORT;

    utSetLogConfig(FALSE, TRUE);	/* use CQ_USERHOMEDIR for logfile */

    cInfo.remotehost = strdup("localhost"); /* default to your own server */

    dspInitData();


#if defined(MINGW)
    /* init windows sockets (version 2.0) */
    rv = WSAStartup(MAKEWORD(2, 0), &wsaData);
    if (rv != 0) {
        fprintf(stderr, "main: WSAStartup failed: %d\n", rv);
        utLog("main: WSAStartup failed: %d\n", rv);
        return 1;
    }
#endif

    /* check options */
    while ((i = getopt(argc, argv, "fmM:s:r:tP:Bug:Sv")) != EOF)
        switch (i)
        {
        case 'B':                 /* Benchmark mode, set recFrameDelay to 0.0 */
            pbSpeed = PB_SPEED_INFINITE;
            recFrameDelay = 0.0;
            break;
        case 'f':
            DSPFSET(DSP_F_FULLSCREEN);
            break;
        case 'g':
            parseGeometry(optarg);
            break;
        case 'm':
            wantMetaList = TRUE;
            break;
        case 'M':
            utStrncpy(cInfo.metaServer, optarg, MAXHOSTNAME);

            break;
        case 's':                 /* [host[:port]] */
            free(cInfo.remotehost);
            cInfo.remotehost = strdup(optarg);
            if (!cInfo.remotehost)
            {
                printf("strdup failed\n");
                exit(1);
            }
            if ((ch = strchr(cInfo.remotehost, ':')) != NULL)
            {
                *ch = 0;
                ch++;
                if ((cInfo.remoteport = atoi(ch)) == 0)
                    cInfo.remoteport = CN_DFLT_PORT;

                /* if no host was specified (only the :port), then set to
                   localhost */
                if (strlen(cInfo.remotehost) == 0)
                {
                    free(cInfo.remotehost);
                    cInfo.remotehost = strdup("localhost");
                }
            }
            else
                cInfo.remoteport = CN_DFLT_PORT;

            serveropt = TRUE;

            break;
        case 'r':
            /* don't want to do this if we've already seen -P */
            if (Context.recmode != RECMODE_PLAYING)
            {
                if (recOpenOutput(optarg, FALSE))
                {			/* we are almost ready... */
                    Context.recmode = RECMODE_STARTING;
                    printf("Recording game to %s...\n", optarg);
                }
                else
                {
                    Context.recmode = RECMODE_OFF;
                    printf("Cannot record game to %s... terminating\n", optarg);
                    exit(1);
                }
            }
            break;

        case 'P':
            recFilename = optarg;
            Context.recmode = RECMODE_PLAYING;
            dosound = FALSE;        /* no sound during playback */
            break;

        case 'u':
            cInfo.tryUDP = FALSE;
            break;

        case 'S':
            dosound = FALSE;
            break;

        case 'v':
            cqDebug++;
            break;

        default:
            printUsage();
            exit(1);
        }

    rndini();		/* initialize random numbers */

    if (!pktInit())
    {
        fprintf(stderr, "pktInit failed, exiting\n");
        utLog("pktInit failed, exiting");
    }
    pktSetSocketFds(cInfo.sock, cInfo.usock);

#ifdef DEBUG_CONFIG
    utLog("%s@%d: main() Reading Configuration files.", __FILE__, __LINE__);
#endif

    if (GetConf(0) == -1)
    {
#ifdef DEBUG_CONFIG
        utLog("%s@%d: main(): GetConf() returned -1.", __FILE__, __LINE__);
#endif
	exit(1);
    }

    Context.updsec = UserConf.UpdatesPerSecond;

    /* For windows, if no -P, -m, or -s was specified, enable -m.  There
     * is no server available under windows, so no need to default to
     * trying to connect to localhost.
     */
#if defined(MINGW)
    if (!serveropt && !wantMetaList && Context.recmode != RECMODE_PLAYING)
        wantMetaList = TRUE;
#endif

    if (Context.recmode == RECMODE_PLAYING)
    {
        if (serveropt || wantMetaList)
            printf("-P option specified.  All other options ignored.\n");

        serveropt = wantMetaList = FALSE;
        printf("Scanning file %s...\n", recFilename);

        /* On windows, over a shared folder in VBox, this will be way slow */
        if (!pbInitReplay(recFilename, &recTotalElapsed))
            exit(1);

        /* now init for real */
        if (!pbInitReplay(recFilename, NULL))
            exit(1);

        Context.unum = -1;       /* stow user number */

        /* turn off annoying beeps */
        UserConf.DoAlarms = FALSE;
        /* turn off hudInfo */
        UserConf.hudInfo = FALSE;
    }

    if (serveropt && wantMetaList)
    {
        printf("-m ignored, since -s was specified\n");
        wantMetaList = FALSE;
    }

    /* load the globals/planets, etc */
    cqiLoadRC(CQI_FILE_CONQINITRC, NULL, 1, 0);

    /* load the main texturesrc file and any user supplied ~/.conquest/ *.trc
       files */
    loadTextureRCFiles();

    /* a parallel universe, it is */
    cbMapLocal();

    /* init and load the sounds */
    if (dosound)
    {
        loadSoundRCFiles();
        cqsInitSound();
    }
    else
        cqsSoundAvailable = FALSE;

#ifdef DEBUG_FLOW
    utLog("%s@%d: main() starting conqinit().", __FILE__, __LINE__);
#endif

    conqinit();			/* machine dependent initialization */
    ibufInit();

#ifdef DEBUG_FLOW
    utLog("%s@%d: main() starting cdinit().", __FILE__, __LINE__);
#endif


    uiGLInit(&argc, argv);

    cqsMusicPlay(cqsFindMusic("intro"), FALSE);

    Context.maxlin = 25;
    Context.maxcol = 80;
    Context.snum = 0;		/* force menu to get a new ship */
    Context.histslot = -1;
    Context.lasttang = Context.lasttdist = 0;

    Context.lasttdist = Context.lasttang = 0;
    Context.lasttarg[0] = 0;
    catchSignals();       /* enable trapping of interesting signals */

    /* which node to start from... */

    if (Context.recmode == RECMODE_PLAYING)
        nPlayBMenuInit();
    else if (wantMetaList)
        nMetaInit();
    else
        nConsvrInit(cInfo.remotehost, cInfo.remoteport);

#ifdef DEBUG_FLOW
    utLog("%s@%d: main() welcoming player.", __FILE__, __LINE__);
#endif

    /* start the fun! */
    glutMainLoop();

    exit(0);

}

void catchSignals(void)
{
#ifdef DEBUG_SIG
    utLog("catchSignals() ENABLED");
#endif
#if !defined(MINGW)
    signal(SIGHUP, (void (*)(int))handleSignal);
    /*  signal(SIGTSTP, SIG_IGN);*/
    signal(SIGTERM, (void (*)(int))handleSignal);
    signal(SIGINT, SIG_IGN);
    signal(SIGQUIT, (void (*)(int))handleSignal);
#endif  /* MINGW */
    return;
}

void handleSignal(int sig)
{

#ifdef DEBUG_SIG
    utLog("handleSignal() got SIG %d", sig);
#endif
#if defined(MINGW)
    utLog("handleSignal() (MINGW) got a signal??? %d", sig);
    exit(1);
#else

    switch(sig)
    {
    case SIGQUIT:
    case SIGINT:
    case SIGTERM:
    case SIGHUP:
        conqend();		/* sends a disconnect packet */
        exit(0);
        break;

    default:
        break;
    }

    catchSignals();	/* reset */
#endif                  /* MINGW */
    return;
}

/*  conqend - machine dependent clean-up */
/*  SYNOPSIS */
/*    conqend */
void conqend(void)
{
    if (cInfo.sock != -1)
        sendCommand(CPCMD_DISCONNECT, 0); /* tell the server */
    recCloseOutput();

    return;

}
